text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1058 @@
+- Start Date: 2016-11-05
+- RFC PR: https://github.com/emberjs/rfcs/pull/176
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Make Ember feel less overwhelming to new users, and make Ember applications
+start faster, by replacing the `Ember` global with a first-class system for
+importing just the parts of the framework you need.
+
+# Motivation
+
+ECMAScript 2015 (also known as ES2015 or ES6) introduced a syntax for importing
+and exporting values from modules. Ember aggressively adopted modules, and if
+you've used Ember before, you're probably familiar with this syntax:
+
+```js
+import Ember from "ember";
+import Analytics from "../mixins/analytics";
+
+export default Ember.Component.extend(Analytics, {
+  // ...
+});
+```
+
+One thing to notice is that the entire Ember framework is imported as a single
+package. Rather than importing `Component` directly, for example, you import
+`Ember` and subclass `Ember.Component`. (And this example still works even if
+you forget the import, because we also create a global variable called
+`Ember`.)
+
+Using Ember via a monolithic package or global object is not ideal for several
+reasons:
+
+* It's overwhelming for learners. There's a giant list of classes and functions
+  with no hints about how they're related. The API documentation reflects this.
+* Experienced developers who don't want all of Ember's features feel like
+  they're adding unnecessary and inescapable bloat to their application.
+* The `Ember` object must be built at boot time, requiring that we ship the
+  entire framework to the browser. This has two major costs:
+  1. Increased download time, particularly noticeable on slower connections.
+  2. Increased parsing/evaluation cost, which still must be paid even when
+     assets are cached. On some browsers/devices, this can far exceed the cost of the
+     download itself.
+* Running multiple versions of Ember on the same page is error-prone.
+
+Defining a public API for importing parts of Ember via JavaScript modules helps
+us lay the groundwork for solving all of these problems.
+
+#### Multiple Apps on a Page
```

Comment:
@nathanhammond I had some trouble feeling confident that I'd grokked your comment as well. 😁 I think you're saying that we don't want to support the case of building "multiple apps" from a single Ember CLI project. That's something I wholeheartedly agree with.

However, I *do* think there are a subset of users who want to incrementally migrate to Ember. For better of worse, that often means replacing chunks of server-rendered pages with Ember apps (until enough of the app is in Ember that you can just move things over wholesale.)

For our adoption goals, I think having a story for doing this is important. I am, of course, not strongly opposed to removing this section from the RFC. But the RFC doesn't say that we will support it; it's simply saying that having modules _unlocks_ that potential. My motivation here is to try to enumerate all of the possibilities that are _impossibilities_ under the current system, as a way to persuade the skeptics that there is real value here and not just academic tilting at windmills.